src/IECoreScene/PrimitiveVariable.cpp

```diff
@@ -121,29 +121,50 @@ namespace
 
 struct Expander
 {
-	typedef DataPtr ReturnType;
 
 	Expander( const std::vector<int> &indices ) : m_indices( indices )
 	{
 	}
 
-	const std::vector<int> &m_indices;
+	template<typename T>
+	DataPtr operator()( const GeometricTypedData<std::vector<T>> *data ) const
+	{
+		typename GeometricTypedData<std::vector<T>>::Ptr result = expand( data );
+		result->setInterpretation( data->getInterpretation() );
+		return result;
+	}
 
 	template<typename T>
-	ReturnType operator() ( T * data )
+	DataPtr operator()( const TypedData<std::vector<T>> *data ) const
+	{
+		return expand( data );
+	}
+
+	DataPtr operator()( const Data *data ) const
 	{
-		const typename T::ValueType &compactValues = data->readable();
+		throw IECore::InvalidArgumentException( "Can only expand VectorTypedData" );
+	}
+
+	private :
 
-		typename T::Ptr result = new T();
-		typename T::ValueType &expandedValues = result->writable();
-		expandedValues.reserve( m_indices.size() );
-		for( const auto &index : m_indices )
+		template<typename DataType>
+		typename DataType::Ptr expand( const DataType *data ) const
 		{
-			expandedValues.push_back( compactValues[index] );
+			const typename DataType::ValueType &compactValues = data->readable();
+
+			typename DataType::Ptr result = new DataType();
+			typename DataType::ValueType &expandedValues = result->writable();
+			expandedValues.reserve( m_indices.size() );
+			for( const auto &index : m_indices )
+			{
+				expandedValues.push_back( compactValues[index] );
+			}
+
+			return result;
 		}
 
-		return result;
-	}
+		const std::vector<int> &m_indices;
```

Comment:
Don and I already discussed the possibility of doing something like this (although I realise you didn't reply to my last email where I proposed the form below, @donboie, is it similar to what you had in mind? ) :

```
struct Expander
{
    template<typename T>
    DataPtr operator()( const PrimitiveVariable &pv, const T *data ) const
    {
    }
    ....
}

PrimitiveAlgo::dispatch( primitiveVariable, Expander() );
```

It doesn't seem unreasonable, but I do wonder how generally applicable it will be; plenty of other operations need access to something like the mesh topology or some other arguments. I don't think we should be adding lots of different dispatch variants to satisfy all such cases.

I also have my eye on the future, where C++14 gives us the ability to dispatch to a generic lambda and C++17 gives us the ability to replace all the overloads with a `if constexpr`. At that point it'll be very natural to bind in additional arguments anyway. So I'd say that if a `PrimitiveAlgo::dispatch()` call is to justify its existence in the short term, it needs to be broadly applicable rather than useful in only a couple of cases.